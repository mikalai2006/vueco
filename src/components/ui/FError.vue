<script setup lang="ts">
defineProps({
  /**
   * String error for field.
   */
  error: {
    type: String,
    default: "",
  },
});
</script>
<template>
  <div
    role="alert"
    class="mt-2 text-sm leading-none text-d-600 dark:text-d-400"
  >
    <slot>
      {{ error }}
    </slot>
  </div>
</template>
