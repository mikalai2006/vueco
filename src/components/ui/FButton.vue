<script lang="ts" setup>
import { computed, ref, type PropType } from "vue";
import { LBtn } from "@/components/LBtn";
import { TypeBtnTag, TypeBtnMode } from "@/components/LBtn";
import { ripple } from "@/directives/ripple";

const vRipple = ripple;

const props = defineProps({
  as: {
    type: String as PropType<TypeBtnTag>,
    default: "button",
    validator(val: TypeBtnTag) {
      return [...Object.values(TypeBtnTag)].includes(val);
    },
  },
  mode: {
    type: String as PropType<TypeBtnMode>,
    default: TypeBtnMode.second,
    validator(val: TypeBtnMode) {
      return [...Object.values(TypeBtnMode)].includes(val);
    },
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  pressed: {
    type: Boolean,
    default: false,
  },
  fill: {
    type: Boolean,
    default: false,
  },
});
defineEmits(["update:pressed"]);

const cssClass = computed(() => {
  let _class =
    "relative py-2 px-4 font-medium rounded-md outline-none cursor-pointer transition focus:ring-2 focus:ring-offset-2 focus:ring-offset-white dark:focus:ring-offset-s-900";
  switch (props.mode as keyof typeof TypeBtnMode) {
    case TypeBtnMode.primary:
      _class +=
        " ring-p-500" +
        (props.fill
          ? " bg-p-500 dark:bg-p-700 text-white hover:bg-p-400 hover:bg-p-600 dark:hover:bg-p-600"
          : " bg-p-500/10 text-p-700 dark:text-p-400 hover:bg-p-500/30");
      break;
    case TypeBtnMode.second:
      _class +=
        " ring-s-500" +
        (props.fill
          ? " bg-s-500 dark:bg-s-700 text-white hover:bg-s-600 dark:hover:bg-s-600"
          : " bg-s-500/10 text-s-700 dark:text-s-400 hover:bg-s-500/30");
      break;
    case TypeBtnMode.danger:
      _class +=
        " ring-d-500" +
        (props.fill
          ? " bg-d-600 dark:bg-d-700 text-white hover:bg-d-700 dark:hover:bg-d-600"
          : " bg-d-600/10 text-d-700 dark:text-d-400 hover:bg-d-500/30");
      break;
    case TypeBtnMode.link:
      _class += " ring-s-500/10 hover:bg-s-500/10";
      break;

    default:
      break;
  }
  if (props.pressed) {
    _class += " shadow-inner";
  }
  return _class;
});

const btn = ref(null);

defineExpose({ $el: btn, el: btn });
</script>

<template>
  <LBtn
    ref="btn"
    v-ripple="{ class: 'bg-black/50' }"
    v-bind="{ disabled, pressed, as }"
    :class="[
      cssClass,
      {
        'cursor-not-allowed disabled:bg-transparent disabled:text-s-500 dark:disabled:bg-transparent dark:disabled:text-s-500':
          disabled,
      },
    ]"
    @click="$emit('update:pressed', !pressed)"
  >
    <span class="relative">
      <slot></slot>
    </span>
  </LBtn>
</template>
