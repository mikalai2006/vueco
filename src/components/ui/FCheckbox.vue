<script lang="ts">
import { defineComponent } from "vue";
export default defineComponent({
  name: "FCheckbox",
  inheritAttrs: false,
});
</script>
<script setup lang="ts">
import { computed, ref } from "vue";
import type { IField } from "@/components/LForm";

import { LInputCheckbox } from "@/components/LInput";
import { LIcon } from "@/components/LIcon";

const props = withDefaults(
  defineProps<{
    /**
     * modelValue - model boolean
     */
    modelValue: boolean;
    error?: string;
    /**
     * Schema for field.
     */
    field: IField;
    required?: boolean;
  }>(),
  {
    modelValue: false,
  }
);

const emit = defineEmits(["update:modelValue", "on-input"]);

const model = computed({
  get: () => props.modelValue,
  set: (val: boolean) => {
    emit("update:modelValue", val);
  },
});

const inputElement = ref<InstanceType<typeof LInputCheckbox> | null>(null);

const onSetFocus = () => {
  const el = inputElement.value?.$el;

  if (el && el?.input) el?.input.focus();
};

const empty = computed(() => model.value != false && model.value != true);

const onReset = () => {
  model.value = false;
  emit("on-input");
  onSetFocus();
};
</script>
<template>
  <div class="relative flex items-start">
    <slot name="before" :onReset="onReset"></slot>
    <div class="flex items-center">
      <LInputCheckbox
        v-bind="$attrs"
        ref="inputElement"
        v-model="model"
        :field="field"
        :error="error"
        :checked="!!model"
        class="mr-2 h-5 w-5 appearance-none outline-none group focus:ring-2 focus:ring-offset-2 border rounded-md focus:ring-offset-white dark:focus:ring-offset-s-800 checked:bg-p-500"
        :class="[
          {
            'border-d-500 dark:border-d-500 focus:ring-d-500 focus:border-d-500 hover:ring-d-500':
              error,
          },
          {
            'border-s-200 dark:border-s-600 focus:ring-p-500 focus:border-p-500 hover:ring-p-500 hover:border-p-500':
              !error,
          },
        ]"
        @change="$emit('on-input')"
      >
        <LIcon
          v-if="model"
          class="bg-transparent absolute top-0 left-0 text-white text-xl"
        >
          M10.97 4.97a.75.75 0 0 1 1.07 1.05l-3.99 4.99a.75.75 0 0
          1-1.08.02L4.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093
          3.473-4.425a.267.267 0 0 1 .02-.022z
        </LIcon>
      </LInputCheckbox>
    </div>
    <div>
      <slot
        name="label"
        :id="field.id"
        :required="required"
        :onSetFocus="onSetFocus"
      ></slot>
      <slot name="after" :onReset="onReset" :empty="empty"></slot>
      <slot name="error"></slot>
      <slot name="description"></slot>
    </div>
  </div>
</template>
