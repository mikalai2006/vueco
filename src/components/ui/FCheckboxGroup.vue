<script lang="ts">
import { defineComponent } from "vue";
export default defineComponent({
  name: "FCheckboxGroup",
  inheritAttrs: false,
});
</script>
<script setup lang="ts">
import { computed, ref } from "vue";
import type { IField } from "@/components/LForm";

import { LInputCheckboxGroup } from "@/components/LInput";
import { LIcon } from "@/components/LIcon";

const props = withDefaults(
  defineProps<{
    /**
     * modelValue - model boolean
     */
    modelValue: Array<string | number>;
    error?: string;
    /**
     * Schema for field.
     */
    field: IField;
    required?: boolean;
  }>(),
  {
    modelValue: () => {
      return [];
    },
  }
);

const emit = defineEmits(["update:modelValue", "on-input"]);

const model = computed({
  get: () => props.modelValue,
  set: (val: Array<string | number>) => {
    emit("update:modelValue", val);
  },
});

const inputElement = ref<InstanceType<typeof LInputCheckboxGroup> | null>(null);

const onSetFocus = () => {
  const el = inputElement.value?.$el;

  if (el && el?.input) el?.input.focus();
};

const onReset = () => {
  model.value = [];
  emit("on-input");
  onSetFocus();
};
</script>
<template>
  <div class="relative flex items-start">
    <slot name="before" :onReset="onReset"></slot>
    <div class="flex items-center">
      <LInputCheckboxGroup
        v-for="item in field.groupvalue"
        :key="`item_${item.id}`"
        v-bind="$attrs"
        ref="inputElement"
        v-model="model"
        :field="field"
        :error="error"
        :checked="model.indexOf(item.value) != -1"
        class="mr-2 h-5 w-5 appearance-none outline-none group focus:ring-2 focus:ring-offset-2 border rounded-md focus:ring-offset-white dark:focus:ring-offset-s-800 checked:bg-p-500"
        :class="[
          {
            'border-d-500 dark:border-d-500 focus:ring-d-500 focus:border-d-500 ':
              error,
          },
          {
            'border-s-200 dark:border-s-600 focus:ring-p-500 focus:border-p-500':
              !error,
          },
        ]"
        @change="$emit('on-input')"
      >
        <LIcon
          v-if="model.indexOf(item.value) !== -1"
          class="bg-transparent absolute top-0 left-0 text-white text-xl"
        >
          M10.97 4.97a.75.75 0 0 1 1.07 1.05l-3.99 4.99a.75.75 0 0
          1-1.08.02L4.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093
          3.473-4.425a.267.267 0 0 1 .02-.022z
        </LIcon>
      </LInputCheckboxGroup>
    </div>
    <div>
      <slot
        name="label"
        :id="field.id"
        :required="required"
        :onSetFocus="onSetFocus"
      ></slot>
      <slot name="after" :onReset="onReset"></slot>
      <slot name="error"></slot>
      <slot name="description"></slot>
    </div>
  </div>
</template>
