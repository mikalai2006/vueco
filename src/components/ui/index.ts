import FError from "./FError.vue";
import FCheckbox from "./FCheckbox.vue";
import FCheckboxGroup from "./FCheckboxGroup.vue";
import FNumber from "./FNumber.vue";
import FTabs from "./FTabs.vue";
import FText from "./FText.vue";
import FTextarea from "./FTextarea.vue";
import FMultiSelect from "./FMultiSelect.vue";

export {
  FError,
  FCheckbox,
  FMultiSelect,
  FCheckboxGroup,
  FNumber,
  FTabs,
  FText,
  FTextarea,
};
