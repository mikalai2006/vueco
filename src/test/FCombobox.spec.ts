import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import { FCombobox } from "@/components/ui/index";
import { ref, defineComponent } from "vue";
import { vi } from "vitest";
import {
  LCombobox,
  LComboboxInput,
  LComboboxOptions,
  LComboboxOption,
} from "@/components/LCombobox";

const testProps = {
  autocomplete: "both",
  id: "friends",
  name: "friends",
  component: "FCombobox",
  keyValue: "title", // ключ для свойства
  keyLabel: "title",
  keyFilter: "title",
  autofocus: true,
  label: "Your friends",
  description: "Choose friends your know",
  placeholder: "Choose friends",
  items: [
    {
      id: 1,
      title: "One",
    },
    {
      id: 2,
      title: "Two",
    },
    {
      id: 3,
      title: "Three",
    },
  ],
  vtype: "array",
  rules: [
    {
      type: "required",
      params: ["This field is required"],
    },
    {
      type: "length",
      params: [1, "Languages cannot be less than 1 language"],
    },
  ],
};
const testComponent = defineComponent({
  components: {
    LCombobox,
    LComboboxInput,
    LComboboxOptions,
    LComboboxOption,
  },
  setup() {
    const model = ref("");
    const field = ref(testProps);
    return {
      model,
      field,
    };
  },
  template: `
  <LCombobox :field="field" v-model="model">
    <LComboboxInput></LComboboxInput>
    <LComboboxOptions #default="{ options, hasOptions }">
        <template v-if="hasOptions">
          <LComboboxOption
            v-for="option of options"
            :key="field.keyValue ? option[field.keyValue] : option"
            :value="option"
            class="group relative overflow-hidden"
            #default="{ selected, hovered, height }"
          ></LComboboxOption>
        </template>
        <template v-else>
          <div class="p-2">None options</div>
        </template>
    </LComboboxOptions>
  </LCombobox>`,
});
const model = ref("");

describe("Combobox with autocomplete", () => {
  it("should be render with default options", () => {
    const wrapper = mount(testComponent);

    expect(wrapper.html()).toBeDefined();
    expect(wrapper.find("input").attributes("aria-expanded")).toBe("false");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should be by press ArrowDown ", async () => {
    const wrapper = mount(testComponent);

    await wrapper.find("input").trigger("click");
    expect(wrapper.find("input").attributes("aria-expanded")).toBe("true");
    await wrapper.find("input").trigger("keydown", { key: "ArrowDown" });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("LCombobox filter", () => {
    // vi.useFakeTimers();
    const wrapper = mount(testComponent);

    wrapper.find("input").trigger("click");
    wrapper.find("input").setValue("Gogo");
    wrapper.vm.$nextTick(() => {
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.findAll("[role='option']").length).toEqual(0);
    });
    // vi.advanceTimersByTime(20000);
  });
});

describe("FCombobox", () => {
  it("renders properly", () => {
    const wrapper = mount(FCombobox, {
      props: { field: testProps, modelValue: model.value },
    });

    expect(wrapper.html()).toBeDefined();
    expect(wrapper.find("input").attributes("aria-expanded")).toBe("false");
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("FCombobox a11y", () => {
  it("autocomplete none", () => {
    const wrapper = mount(FCombobox, {
      props: {
        field: { ...testProps, autocomplete: "none" },
        modelValue: model.value,
      },
    });

    expect(wrapper.html()).toBeDefined();
    expect(wrapper.find("[role='combobox']")).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("autocomplete both", () => {
    const wrapper = mount(FCombobox, {
      props: {
        field: { ...testProps, autocomplete: "both" },
        modelValue: model.value,
      },
    });

    expect(wrapper.html()).toBeDefined();
    expect(wrapper.find("[role='combobox']")).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("role=listbox", () => {
    const wrapper = mount(FCombobox, {
      props: { field: testProps, modelValue: model.value },
    });

    expect(wrapper.find("[role='listbox']").attributes("id")).toBe(
      `${testProps.id}-listbox`
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("test aria-expanded", async () => {
    const wrapper = mount(FCombobox, {
      props: { field: testProps, modelValue: model.value },
    });

    await wrapper.find("input").trigger("click");
    expect(wrapper.find("input").attributes("aria-expanded")).toBe("true");
  });

  it("test aria-activedescendant(first)", async () => {
    const wrapper = mount(FCombobox, {
      props: { field: testProps, modelValue: model.value },
    });

    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown", { key: "ArrowDown" });
    const activedescendant = wrapper
      .find("input")
      .attributes("aria-activedescendant");
    expect(wrapper.findAll("[role='option']")[0].attributes("id")).toBe(
      activedescendant
    );
  });
  it("test aria-activedescendant(last)", async () => {
    const wrapper = mount(FCombobox, {
      props: { field: testProps, modelValue: model.value },
    });

    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown", { key: "ArrowUp" });
    const activedescendant = wrapper
      .find("input")
      .attributes("aria-activedescendant");
    const allOptions = wrapper.findAll("[role='option']");
    expect(allOptions[allOptions.length - 1].attributes("id")).toBe(
      activedescendant
    );
    expect(wrapper.html()).toMatchSnapshot();
  });
  // it("load items", () => {
  //   const wrapper = mount(FCombobox, {
  //     props: { field: testProps, modelValue: model },
  //     slots: {

  //     }
  //   });
  //   expect(wrapper.html()).toBeDefined();
  //   expect(wrapper.html()).toMatchSnapshot();
  // });
});
