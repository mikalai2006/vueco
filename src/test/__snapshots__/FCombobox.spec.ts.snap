// Vitest Snapshot v1

exports[`Combobox with autocomplete > LCombobox filter 1`] = `
"<div><input id=\\"friends\\" type=\\"text\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" placeholder=\\"Choose friends\\" aria-autocomplete=\\"both\\" aria-expanded=\\"true\\" aria-controls=\\"friends-listbox\\">
  <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\">
    <div class=\\"p-2\\">None options</div>
  </ul>
</div>"
`;

exports[`Combobox with autocomplete > should be by press ArrowDown  1`] = `
"<div><input id=\\"friends\\" type=\\"text\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" placeholder=\\"Choose friends\\" aria-autocomplete=\\"both\\" aria-expanded=\\"true\\" aria-controls=\\"friends-listbox\\" aria-activedescendant=\\"friends-option-4\\">
  <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\">
    <li id=\\"friends-option-4\\" role=\\"option\\" aria-disabled=\\"false\\" tabindex=\\"-1\\" class=\\"group relative overflow-hidden\\">One</li>
    <li id=\\"friends-option-5\\" role=\\"option\\" aria-disabled=\\"false\\" tabindex=\\"-1\\" class=\\"group relative overflow-hidden\\">Two</li>
    <li id=\\"friends-option-6\\" role=\\"option\\" aria-disabled=\\"false\\" tabindex=\\"-1\\" class=\\"group relative overflow-hidden\\">Three</li>
  </ul>
</div>"
`;

exports[`Combobox with autocomplete > should be render with default options 1`] = `
"<div><input id=\\"friends\\" type=\\"text\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" placeholder=\\"Choose friends\\" aria-autocomplete=\\"both\\" aria-expanded=\\"false\\" aria-controls=\\"friends-listbox\\">
  <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\">
    <div class=\\"p-2\\">None options</div>
  </ul>
</div>"
`;

exports[`FCombobox > renders properly 1`] = `
"<div class=\\"flex items-center cursor-pointer rounded-md border text-base transition w-full appearance-none outline-none focus-within:ring-1 hover:ring-1  bg-white dark:bg-s-900 border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500 hover:ring-p-500 hover:border-p-500\\">
  <div class=\\"relative flex-auto\\">
    <div class=\\"outline-none flex space-x-2 items-center w-full transition\\">
      <div class=\\"flex-auto flex flex-wrap m-1\\">
        <!--v-if--><input id=\\"friends\\" type=\\"text\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" placeholder=\\"Choose friends\\" aria-autocomplete=\\"both\\" aria-expanded=\\"false\\" aria-controls=\\"friends-listbox\\" autocomplete=\\"off\\" class=\\"w-full px-2 py-1 cursor-text outline-none bg-transparent placeholder-s-400\\">
      </div>
      <div class=\\"py-2 pr-2\\"><svg class=\\"fill-current transition-transform\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\">
          <path d=\\"M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z\\"></path>
        </svg></div>
      <div class=\\"absolute right-8 top-0 h-full flex items-center\\"></div>
    </div>
    <div class=\\"z-10 mt-1.5 transition overflow-hidden shadow-2xl outline-none border border-s-200 dark:border-s-700 rounded-md absolute top-full left-0 right-0 bg-white dark:bg-s-800 hidden border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500\\" style=\\"display: none;\\">
      <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\" class=\\"max-h-64 b-scroll overflow-y-auto\\">
        <div class=\\"p-2\\">None options</div>
      </ul>
    </div>
  </div>
</div>
<!--v-if-->"
`;

exports[`FCombobox a11y > autocomplete both 1`] = `
"<div class=\\"flex items-center cursor-pointer rounded-md border text-base transition w-full appearance-none outline-none focus-within:ring-1 hover:ring-1  bg-white dark:bg-s-900 border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500 hover:ring-p-500 hover:border-p-500\\">
  <div class=\\"relative flex-auto\\">
    <div class=\\"outline-none flex space-x-2 items-center w-full transition\\">
      <div class=\\"flex-auto flex flex-wrap m-1\\">
        <!--v-if--><input id=\\"friends\\" type=\\"text\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" placeholder=\\"Choose friends\\" aria-autocomplete=\\"both\\" aria-expanded=\\"false\\" aria-controls=\\"friends-listbox\\" autocomplete=\\"off\\" class=\\"w-full px-2 py-1 cursor-text outline-none bg-transparent placeholder-s-400\\">
      </div>
      <div class=\\"py-2 pr-2\\"><svg class=\\"fill-current transition-transform\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\">
          <path d=\\"M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z\\"></path>
        </svg></div>
      <div class=\\"absolute right-8 top-0 h-full flex items-center\\"></div>
    </div>
    <div class=\\"z-10 mt-1.5 transition overflow-hidden shadow-2xl outline-none border border-s-200 dark:border-s-700 rounded-md absolute top-full left-0 right-0 bg-white dark:bg-s-800 hidden border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500\\" style=\\"display: none;\\">
      <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\" class=\\"max-h-64 b-scroll overflow-y-auto\\">
        <div class=\\"p-2\\">None options</div>
      </ul>
    </div>
  </div>
</div>
<!--v-if-->"
`;

exports[`FCombobox a11y > autocomplete none 1`] = `
"<div class=\\"flex items-center cursor-pointer rounded-md border text-base transition w-full appearance-none outline-none focus-within:ring-1 hover:ring-1  bg-white dark:bg-s-900 border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500 hover:ring-p-500 hover:border-p-500\\">
  <div class=\\"relative flex-auto\\">
    <div class=\\"outline-none flex space-x-2 items-center w-full transition\\">
      <div class=\\"flex-auto flex flex-wrap m-1\\">
        <div id=\\"friends\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" aria-labelledby=\\"friends_label\\" aria-expanded=\\"false\\" aria-controls=\\"friends-listbox\\" aria-activedescendant=\\"\\" autocomplete=\\"off\\" class=\\"w-full px-2 py-1 outline-none bg-transparent\\">
          <div class=\\"text-s-400\\">Choose friends</div>
        </div>
      </div>
      <div class=\\"py-2 pr-2\\"><svg class=\\"fill-current transition-transform\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\">
          <path d=\\"M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z\\"></path>
        </svg></div>
      <div class=\\"absolute right-8 top-0 h-full flex items-center\\"></div>
    </div>
    <div class=\\"z-10 mt-1.5 transition overflow-hidden shadow-2xl outline-none border border-s-200 dark:border-s-700 rounded-md absolute top-full left-0 right-0 bg-white dark:bg-s-800 hidden border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500\\" style=\\"display: none;\\">
      <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\" class=\\"max-h-64 b-scroll overflow-y-auto\\">
        <div class=\\"p-2\\">None options</div>
      </ul>
    </div>
  </div>
</div>
<!--v-if-->"
`;

exports[`FCombobox a11y > role=listbox 1`] = `
"<div class=\\"flex items-center cursor-pointer rounded-md border text-base transition w-full appearance-none outline-none focus-within:ring-1 hover:ring-1  bg-white dark:bg-s-900 border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500 hover:ring-p-500 hover:border-p-500\\">
  <div class=\\"relative flex-auto\\">
    <div class=\\"outline-none flex space-x-2 items-center w-full transition\\">
      <div class=\\"flex-auto flex flex-wrap m-1\\">
        <!--v-if--><input id=\\"friends\\" type=\\"text\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" placeholder=\\"Choose friends\\" aria-autocomplete=\\"both\\" aria-expanded=\\"false\\" aria-controls=\\"friends-listbox\\" autocomplete=\\"off\\" class=\\"w-full px-2 py-1 cursor-text outline-none bg-transparent placeholder-s-400\\">
      </div>
      <div class=\\"py-2 pr-2\\"><svg class=\\"fill-current transition-transform\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\">
          <path d=\\"M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z\\"></path>
        </svg></div>
      <div class=\\"absolute right-8 top-0 h-full flex items-center\\"></div>
    </div>
    <div class=\\"z-10 mt-1.5 transition overflow-hidden shadow-2xl outline-none border border-s-200 dark:border-s-700 rounded-md absolute top-full left-0 right-0 bg-white dark:bg-s-800 hidden border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500\\" style=\\"display: none;\\">
      <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\" class=\\"max-h-64 b-scroll overflow-y-auto\\">
        <div class=\\"p-2\\">None options</div>
      </ul>
    </div>
  </div>
</div>
<!--v-if-->"
`;

exports[`FCombobox a11y > test aria-activedescendant(last) 1`] = `
"<div class=\\"flex items-center cursor-pointer rounded-md border text-base transition w-full appearance-none outline-none focus-within:ring-1 hover:ring-1  bg-white dark:bg-s-900 border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500 hover:ring-p-500 hover:border-p-500\\">
  <div class=\\"relative flex-auto\\">
    <div class=\\"outline-none flex space-x-2 items-center w-full transition\\">
      <div class=\\"flex-auto flex flex-wrap m-1\\">
        <!--v-if--><input id=\\"friends\\" type=\\"text\\" name=\\"friends\\" role=\\"combobox\\" tabindex=\\"0\\" placeholder=\\"Choose friends\\" aria-autocomplete=\\"both\\" aria-expanded=\\"true\\" aria-controls=\\"friends-listbox\\" autocomplete=\\"off\\" class=\\"w-full px-2 py-1 cursor-text outline-none bg-transparent placeholder-s-400\\" aria-activedescendant=\\"friends-option-27\\">
      </div>
      <div class=\\"py-2 pr-2\\"><svg class=\\"fill-current transition-transform rotate-180\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\">
          <path d=\\"M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z\\"></path>
        </svg></div>
      <div class=\\"absolute right-8 top-0 h-full flex items-center\\"></div>
    </div>
    <div class=\\"z-10 mt-1.5 transition overflow-hidden shadow-2xl outline-none border border-s-200 dark:border-s-700 rounded-md absolute top-full left-0 right-0 bg-white dark:bg-s-800 block border-s-200 dark:border-s-600 focus-within:ring-p-500 focus-within:border-p-500\\" style=\\"\\">
      <ul id=\\"friends-listbox\\" role=\\"listbox\\" aria-labelledby=\\"friends\\" class=\\"max-h-64 b-scroll overflow-y-auto\\">
        <li id=\\"friends-option-25\\" role=\\"option\\" aria-disabled=\\"false\\" tabindex=\\"-1\\" class=\\"group relative overflow-hidden\\">
          <div class=\\"flex p-2 items-center cursor-pointer focus:outline-none bg-transparent text-black dark:text-white group-focus:bg-p-100 group-hover:bg-p-100 dark:group-hover:bg-p-900 dark:group-focus:bg-p-900\\">
            <div class=\\"flex items-center h-6 w-6 mr-2\\">
              <!--v-if-->
            </div> One
          </div>
        </li>
        <li id=\\"friends-option-26\\" role=\\"option\\" aria-disabled=\\"false\\" tabindex=\\"-1\\" class=\\"group relative overflow-hidden\\">
          <div class=\\"flex p-2 items-center cursor-pointer focus:outline-none bg-transparent text-black dark:text-white group-focus:bg-p-100 group-hover:bg-p-100 dark:group-hover:bg-p-900 dark:group-focus:bg-p-900\\">
            <div class=\\"flex items-center h-6 w-6 mr-2\\">
              <!--v-if-->
            </div> Two
          </div>
        </li>
        <li id=\\"friends-option-27\\" role=\\"option\\" aria-disabled=\\"false\\" tabindex=\\"-1\\" class=\\"group relative overflow-hidden\\">
          <div class=\\"flex p-2 items-center cursor-pointer focus:outline-none bg-p-200 dark:bg-p-800 text-black dark:text-white group-focus:bg-p-200 group-hover:bg-p-200 dark:group-hover:bg-p-800 dark:group-focus:bg-p-800\\">
            <div class=\\"flex items-center h-6 w-6 mr-2\\">
              <!--v-if-->
            </div> Three
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
<!--v-if-->"
`;
